api/internalutils/stream/stream.go

```diff
@@ -372,3 +372,97 @@ func RateLimit[T any](stream Stream[T], wait func() error) Stream[T] {
 		wait:  wait,
 	}
 }
+
+// compareStreams is an adapter for merging two streams based on a `compare` function, it will get the next items from both streams and yield the result of streamA if the compare function returns true,
+// or the result of streamB if false.
+// The streams can be of different types, however, the resulting merged stream must be of one type.
+type compareStreams[T, U, V any] struct {
+	streamA Stream[T]
+	streamB Stream[U]
+	itemA   T
+	itemB   U
+	// hasItemA keeps track of whether we have an item available from streamA.
+	// We use this flag instead of just checking whether itemA is nil in order to ensure that this adapter also works with non-nullable types.
+	hasItemA bool
+	// hasItemB keeps track of whether we have an item available from streamB.
+	hasItemB bool
+	compare  func(a T, b U) bool
```

Comment:
I would rename `compare` to `less`, just like the [`sort`](https://pkg.go.dev/sort) package does.